src/service/folders.rs

```diff
@@ -13,186 +13,159 @@
+    PermittedFoldersMissingUser,
+
+    /// An error occured trying to get the list of permitted folders in
+    /// enterprise mode using the validator.
+    #[error("InfraError# {0}")]
+    PermittedFoldersValidator(String),
+}
+
 #[tracing::instrument(skip(folder))]
 pub async fn save_folder(
     org_id: &str,
     mut folder: Folder,
     is_internal: bool,
-) -> Result<HttpResponse, Error> {
+) -> Result<Folder, FolderError> {
     folder.name = folder.name.trim().to_string();
     if folder.name.is_empty() {
-        return Ok(
-            HttpResponse::InternalServerError().json(MetaHttpResponse::message(
-                http::StatusCode::BAD_REQUEST.into(),
-                "folder name not allow empty".to_string(),
-            )),
-        );
+        return Err(FolderError::MissingName);
     }
 
     if !is_internal && folder.folder_id == DEFAULT_FOLDER {
-        return Ok(
-            HttpResponse::InternalServerError().json(MetaHttpResponse::message(
-                http::StatusCode::BAD_REQUEST.into(),
-                "can't update default Dashboard folder".to_string(),
-            )),
-        );
+        return Err(FolderError::UpdateDefaultFolder);
     }
+
     if folder.folder_id != DEFAULT_FOLDER {
         folder.folder_id = ider::generate();
     }
 
-    match db::folders::put(org_id, folder).await {
-        Ok(folder) => {
-            set_ownership(org_id, "folders", Authz::new(&folder.folder_id)).await;
-            Ok(HttpResponse::Ok().json(folder))
-        }
-        Err(error) => Ok(
-            HttpResponse::InternalServerError().json(MetaHttpResponse::message(
-                http::StatusCode::INTERNAL_SERVER_ERROR.into(),
-                error.to_string(),
-            )),
-        ),
-    }
+    let folder = table::folders::put(org_id, folder).await?;
+    set_ownership(org_id, "folders", Authz::new(&folder.folder_id)).await;
+    Ok(folder)
 }
 
 #[tracing::instrument(skip(folder))]
 pub async fn update_folder(
     org_id: &str,
     folder_id: &str,
     mut folder: Folder,
-) -> Result<HttpResponse, Error> {
+) -> Result<Folder, FolderError> {
     if folder_id.eq(DEFAULT_FOLDER) {
-        return Ok(
-            HttpResponse::InternalServerError().json(MetaHttpResponse::message(
-                http::StatusCode::BAD_REQUEST.into(),
-                "can't update default Dashboard folder".to_string(),
-            )),
-        );
-    }
-    folder.folder_id = folder_id.to_string();
-
-    if let Err(error) = db::folders::put(org_id, folder).await {
-        return Ok(
-            HttpResponse::InternalServerError().json(MetaHttpResponse::message(
-                http::StatusCode::INTERNAL_SERVER_ERROR.into(),
-                error.to_string(),
-            )),
-        );
+        return Err(FolderError::UpdateDefaultFolder);
     }
 
-    Ok(HttpResponse::Ok().json(MetaHttpResponse::message(
-        http::StatusCode::OK.into(),
-        "Dashboard folder updated".to_string(),
-    )))
```

Comment:
On success the old `update_folder` function actually returned a 200 (OK) response with only a success message in the body, but no folder structure in the body.